modules/play-json/src/main/scala/PlayJsonModule.scala

```diff
@@ -0,0 +1,170 @@
+      // Shamelessly copied from the play-json library
+      case (JsSuccess(f, _), JsSuccess(a, _)) => JsSuccess(f(a))
+      case (JsError(e1), JsError(e2))         => JsError(JsError.merge(e1, e2))
+      case (JsError(e), _)                    => JsError(e)
+      case (_, JsError(e))                    => JsError(e)
+
+    }
+  }
+
+  // This is needed to allow undefined optional fields to be treated as `None`.
+  final private def undefinedAsNull[A](field: String, r: Reads[A]): Reads[A] = Reads { json =>
+    ((json \ field) match {
+      case JsDefined(v) => r.reads(v)
+      case _            => r.reads(JsNull)
+    }).repath(JsPath \ field)
+  }
+
+  final private val labellingSeed =
+    new (FSchema[R.Prim, R.SumTermId, R.ProductTermId, ?] ~> LabelledSchema) {
+
+      def apply[A](fSchema: FSchema[R.Prim, R.SumTermId, R.ProductTermId, A]): LabelledSchema[A] =
+        (false, fSchema)
+    }
+
+  implicit final def reads(
+    implicit primNT: R.Prim ~> Reads,
+    branchLabel: R.SumTermId <~< String,
+    fieldLabel: R.ProductTermId <~< String
+  ): RInterpreter[Reads] =
+    new Interpreter[R.Prim, R.SumTermId, R.ProductTermId, Reads] {
+
+      val alg = new (HEnvT[Boolean, RSchema, Reads, ?] ~> Reads) {
+
+        def apply[A](schema: HEnvT[Boolean, RSchema, Reads, A]): Reads[A] = schema.fa match {
+          case One() =>
+            Reads {
+              case JsNull => JsSuccess(())
+              case _      => JsError(Seq(JsPath -> Seq(JsonValidationError("error.expected.null"))))
+            }
+          case :+:(left, right) =>
+            Reads(
+              json =>
+                left
+                  .reads(json)
+                  .fold(
+                    el =>
+                      right.reads(json).map(\/-.apply) match {
+                        case JsError(er) => JsError(JsError.merge(el, er))
+                        case x           => x
+                      },
+                    a => JsSuccess(-\/(a))
+                  )
+            )
+          case p: :*:[Reads, a, b, R.Prim, R.SumTermId, R.ProductTermId] =>
+            if (schema.ask)
+              p.left.and(p.right)((x: a, y: b) => (x, y))
+            else
+              (JsPath \ "_1")
+                .read(p.left)
+                .and((JsPath \ "_2").read(p.right))((x: a, y: b) => (x, y))
+          case PrimSchema(p)       => primNT(p)
+          case SumTerm(id, schema) => undefinedAsNull(branchLabel(id), schema)
+          case u: Union[R.Prim, R.SumTermId, R.ProductTermId, Reads, A, a] =>
+            u.choices.map(u.iso.get)
+          case ProductTerm(id, schema) => undefinedAsNull(fieldLabel(id), schema)
+          case r: Record[R.Prim, R.SumTermId, R.ProductTermId, Reads, A, a] =>
+            r.fields.map(r.iso.get)
+          case SeqSchema(elem) =>
+            Reads {
+              case JsArray(elems) =>
+                elems.toList.traverse(elem.reads _)
```

Comment:
Why don’t we need to wrap with `undefinedAsNull` here?

Reply:
![You're right!](https://media.giphy.com/media/3oFzmkkwfOGlzZ0gxi/giphy.gif)

We should, my bad.

Reply:
Anyway, I guess that if we add `Optional` as an AST node we won’t need `undefinedAsNull` anymore :)

Reply:
I won't post this GIF twice, but "le coeur y est" :)